service/storage_proxy.cc

```diff
@@ -3465,9 +3465,12 @@ class abstract_read_executor : public enable_shared_from_this<abstract_read_exec
             return _proxy->query_mutations_locally(_schema, cmd, _partition_range, timeout, _trace_state);
         } else {
             tracing::trace(_trace_state, "read_mutation_data: sending a message to /{}", ep);
-            return _proxy->_messaging.send_read_mutation_data(netw::messaging_service::msg_addr{ep, 0}, timeout, *cmd, _partition_range).then([this, ep](rpc::tuple<reconcilable_result, rpc::optional<cache_temperature>> result_and_hit_rate) {
+            clock_type::time_point sent_at = clock_type::now();
+            _proxy->mark_sent(ep, sent_at);
+            return _proxy->_messaging.send_read_mutation_data(netw::messaging_service::msg_addr{ep, 0}, timeout, *cmd, _partition_range).then([this, ep, sent_at](rpc::tuple<reconcilable_result, rpc::optional<cache_temperature>> result_and_hit_rate) {
                 auto&& [result, hit_rate] = result_and_hit_rate;
                 tracing::trace(_trace_state, "read_mutation_data: got response from /{}", ep);
+                _proxy->mark_responded(ep, clock_type::now() - sent_at);
```

Comment:
I'm still trying to figure out the theory of what it is you're trying to detect - a node being down or being slow :-(
To detect down nodes I'd expect you should remember *when* the node last responded - i.e., now() - not the latency now() - sent_at as you do now... But even with now(), it's not trivial how exactly you detect a node being down. Maybe you can write some of the "algorithm" in the commit message.

Reply:
Right, I think it's an artifact from the time when this series was both about detecting dead and overloaded nodes. This failure detector shouldn't be interested in how much time it took for a node to respond, as long as it responded at all. I'll go over this code one more time.